### DISCUSSION TITLE
Additional Methods in ActionHook to retrieve Settings

### DISCUSSION THREAD

**Discussion by kschmiechen** [SchweizerischeBundesbahnen/ch.sbb.polarion.extension.interceptor-manager#70] [Ideas]
Hi SBB Team,

I want to suggest some methods in ActionHook to make it easier to retrieve the hook settings.

I analyzed the settings of my existing hooks, had a look at what is already implemented in the new interceptor manager and also considered the methods available for workflow functions and conditions.

The basic methods would be
```
Boolean getSettingsValueAsBoolean(String settingsId)
Integer getSettingsValueAsInt(String settingsId)
String getSettingsValueAsString(String settingsId)
List<String> getSettingsValueAsList(String settingsId)
Map<String, String> getSettingsValuesWithPrefix(String settingsPrefix)
boolean isCommaSeparatedHasItem(String itemToCheck, String settingsId)
```
These work like the methods available for functions and conditions. The methods return `null` if the setting is not available.

In the code of the interceptor manager, I already saw `isCommaSeparatedSettingsHasItem(String itemToCheck, String settingsName, String projectId)` to filter the setting by project ID and also supporting the wildcard `*` in the saved settings.

I want to take this idea one step further and suggest
```
Boolean getSettingsValueAsBoolean(List<String> settingsId)
Integer getSettingsValueAsInt(List<String> settingsId)
String getSettingsValueAsString(List<String> settingsId)
List<String> getSettingsValueAsList(List<String> settingsId)
Map<String, String> getSettingsValuesWithPrefix(List<String> settingsPrefix)
Map<String, String> getSettingsValuesWithPrefix(List<String> settingsPrefix, boolean getAll)
boolean isCommaSeparatedHasItem(String itemToCheck, List<String> settingsId)
```
where every list element of `settingsId` or `settingsPrefix` is one part of the settingId stored in the dot notation.
Example: `List.of("projectId", "typeId", "fieldId")` matches `projectId.typeId.fieldId` in the hook settings.
Moreover, it matches the following wildcard combinations but always returns the first one in this order:
```
projectId.typeId.fieldId=...
projectId.typeId.*      =...
projectId.*     .fieldId=...
projectId.*     .*      =...
*        .typeId.fieldId=...
*        .typeId.*      =...
*        .*     .fieldId=...
*        .*     .*      =...

```
The method `getSettingsValuesWithPrefix(List<String> settingsPrefix, boolean getAll)` has the special behavior if `getAll == true` that it returns not just the entries with the highest priority according to the above order, but all entries matching the prefix as long as they don't overwrite higher priority settings.  
Example (both are added):  
```
projectId.typeId.fieldId.something
projectId.typeId.      *.somethingElse
```
Example (only first is added since the second would overwrite the first entry in the map):  
```
projectId.typeId.fieldId.something
projectId.typeId.      *.something
```

Since this approach is generic, it does not matter how each hook implements it's parameters. Based on my existing hooks, I identified the following examples which I think all work with the above approach:
```
projectId.typeId.fieldId[.columnId]=linkRoleId, targetTypeId[, targetProjectId]
```
```
projectIds=projectId1, projectId2
```
```
projectId1.typeId1=true
projectId1.typeId2=true
projectId2.typeId2=true
```
```
regex.projectId.typeId.fieldId=expression
message.projectId.typeId.fieldId=message
```
```
projectId.typeId.fieldId=columnId
```

**Comment by Jumas** [comment]
Hi Kevin,

We've just merged the changes you requested. There were slight modifications to the method signatures, but everything should work as you expected.

Please let us know if there's anything we missed. Thanks for your suggestion.

**Reply by kschmiechen** [reply]
cool, I'll give it a try once you release the new version
